content/docs/02-guides/03-llama-3_1.mdx

```diff
@@ -0,0 +1,425 @@
+
+Let's look at how you can build a chatbot with [Next.js](https://nextjs.org), Vercel AI SDK, and Llama 3.1 (via [Groq](https://groq.com/)):
+
+```tsx filename="app/api/chat/route.ts"
+import { convertToCoreMessages, streamText } from 'ai';
+import { createOpenAI as createGroq } from '@ai-sdk/openai';
+
+const groq = createGroq({
+  baseURL: 'https://api.groq.com/openai/v1',
+  apiKey: process.env.GROQ_API_KEY,
+});
+
+// Allow streaming responses up to 30 seconds
+export const maxDuration = 30;
+
+export async function POST(req: Request) {
+  const { messages } = await req.json();
+
+  const result = await streamText({
+    model: groq('llama-3.1-70b-versatile'),
+    system: 'You are a helpful assistant.',
+    messages: convertToCoreMessages(messages),
+  });
+
+  return result.toAIStreamResponse();
+}
+```
+
+```tsx filename="app/page.tsx"
+'use client';
+
+import { useChat } from 'ai/react';
+
+export default function Page() {
+  const { messages, input, handleInputChange, handleSubmit } = useChat();
+
+  return (
+    <>
+      {messages.map(message => (
+        <div key={message.id}>
+          {message.role === 'user' ? 'User: ' : 'AI: '}
+          {message.content}
+        </div>
+      ))}
+      <form onSubmit={handleSubmit}>
+        <input name="prompt" value={input} onChange={handleInputChange} />
+        <button type="submit">Submit</button>
+      </form>
+    </>
+  );
+}
+```
+
+The useChat hook on your root page (`app/page.tsx`) will make a request to your AI provider endpoint (`app/api/chat/route.ts`) whenever the user submits a message. The messages are then streamed back in real-time and displayed in the chat UI.
+
+This enables a seamless chat experience where the user can see the AI response as soon as it is available, without having to wait for the entire response to be received.
+
+### Going Beyond Text
+
+The Vercel AI SDK's React Server Components (RSC) API enables you to create rich, interactive interfaces that go beyond simple text generation. With the `streamUI` function, you can dynamically stream React components from the server to the client.
+
+Let's dive into how you can leverage tools with [AI SDK RSC](/docs/ai-sdk-rsc/overview) to build a generative user interface with Next.js (App Router).
+
+First, create a Server Action.
+
+```tsx filename="app/actions.tsx"
+'use server';
+
+import { streamUI, tool } from 'ai/rsc';
+import { createOpenAI as createGroq } from '@ai-sdk/openai';
+import { z } from 'zod';
+
+const groq = createGroq({
+  baseURL: 'https://api.groq.com/openai/v1',
+  apiKey: process.env.GROQ_API_KEY,
+});
+
+export async function streamComponent() {
+  const result = await streamUI({
```

Comment:
Do we want to talk about rate limiting or protecting this endpoint? Or link out to other content?

Reply:
hm - I don't think it's in the scope of this guide. Can resolve for now and then fast follow with a section in next steps?